server/src/test/java/org/elasticsearch/tasks/CancelTaskResponseTests.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.TaskOperationFailure;
+import org.elasticsearch.action.admin.cluster.node.tasks.cancel.CancelTasksResponse;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.test.AbstractXContentTestCase;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+
+public class CancelTaskResponseTests  extends AbstractXContentTestCase<CancelTasksResponse> {
+
+    @Override
+    protected CancelTasksResponse createTestInstance() {
+        List<TaskInfo> tasks = new ArrayList<>();
+        for (int i = 0; i < randomInt(10); i++) {
+            tasks.add(TaskInfoTests.randomTaskInfo());
+        }
+        List<TaskOperationFailure> taskFailures = new ArrayList<>();
+        for (int i = 0; i < randomInt(5); i++) {
+            taskFailures.add(new TaskOperationFailure(
+                randomAlphaOfLength(5), randomNonNegativeLong(), new IllegalStateException("message")));
+        }
+        return new CancelTasksResponse(tasks, taskFailures, Collections.singletonList(new FailedNodeException("", "message", null)));
+    }
+
+    @Override
+    protected CancelTasksResponse doParseInstance(XContentParser parser) {
+        return CancelTasksResponse.fromXContent(parser);
+    }
+
+    @Override
+    protected boolean supportsUnknownFields() {
+        return false;
```

Comment:
this should be set to true so we test forward compatibility. Did you set it to false because you encountered problems with this?

Reply:
@javanna I did have problems with this value set to true, (https://gist.github.com/pcsanwald/748164dc568c672d51b01dd28c7ab6c6)[here's a gist] with the error I saw.

I used the ListTask stuff as kind of a template for my changes, and I noticed we also set this value to false in `ListTasksResponseTest.java`, so I figured I'd need to do the same thing here. Is there a problem in both places? I'm not sure I fully understand the reason behind the failure, actually.

Reply:
the gist looks about the xcontent equivalence checks more than unknown fields. That is kind of expected whenever the response holds an exception, as we parse any exception back into a generic ElasticsearchException, which has a different toXContent output compared to the original exception.

Reply:
Did a bit more debugging here: The problem, which also exists in `ListTasksResponseTests`, is that the `headers` field isn't being parsed correctly. I initially thought that this was because we were not overriding `getRandomFieldsExcludeFilter` (which `TaskInfoTest` does do) on either `ListTasksResponseTests` or `CancelTasksResponseTests`, but, I still am getting a parse error on the headers field.

The error seems to happen most often when the headers field is an empty map (which obv happens in a subset of the cases because it's random). Any thoughts on where the problem might be? I'm still not familiar enough with this code to triangulate, although I'm getting there :).

Reply:
I had a look and reworked a bit tests for list tasks here: #31001 . That should clarify what should and should not work in these tests, your cancel response tests should end up looking very similar. headers and status need to be both excluded, then we can enable the insertion of random fields by setting supportsUnknownFields to true. Also, xcontent equivalence works only when there are no failures, hence I split the two scenarios (with and without failures).

Reply:
That doesn't really answer your question on the empty map problem that you saw, I don't quite follow on the empty map issue but I think it may have to do with the exclusion predicate. I have `return field -> field.endsWith("status") || field.endsWith("headers");`. Note the `endsWith` here compared to the equals in `TaskInfoTests`, that is because such fields are at the top-level for `TaskInfo` but deeper in the ListTasksResponse or CancelTasksResponse output.